test/Microsoft.AspNet.Mvc.Core.Test/Routing/InnerAttributeRouteTest.cs

```diff
@@ -1464,6 +1464,62 @@ public static IEnumerable<object[]> OptionalParamValues
         }
 
         [Fact]
+        public async Task AttributeRoute_ReplacesExistingRouteValues_IfNotNull()
+        {
+            // Arrange
+            var router = new Mock<IRouter>();
+            router
+                .Setup(r => r.RouteAsync(It.IsAny<RouteContext>()))
+                .Callback<RouteContext>((c) =>
+                {
+                    c.IsHandled = true;
+                })
+                .Returns(Task.FromResult(true));
+
+            var entry = CreateMatchingEntry(router.Object, "Foo/{*path}", order: 0);
+            var route = CreateAttributeRoute(router.Object, entry);
+
+            var context = CreateRouteContext("/Foo/Bar");
+
+            var originalRouteData = context.RouteData;
+            originalRouteData.Values.Add("path", "default");
+
+            // Act
+            await route.RouteAsync(context);
+
+            // Assert
+            Assert.Equal("Bar", context.RouteData.Values["path"]);
+        }
+
+        [Fact]
```

Comment:
Routing creates a defensive copy at each level so that changes that aren't persisted don't dirty something else on the fallback path. That's what's on the whiteboard next to you, you silly goose! :goat: